interpreter/valid/valid.ml

```diff
@@ -326,32 +326,35 @@ let rec check_instr (c : context) (e : instr) (s : infer_result_type) : op_type
     [value_type_of_index_type it; RefType t] --> []
 
   | TableSize x ->
-    let _tt = table c x in
-    [] --> [NumType I32Type]
+    let TableType (_lim, it, _t) = table c x in
+    [] --> [value_type_of_index_type it]
 
   | TableGrow x ->
-    let TableType (_lim, _it, t) = table c x in
-    [RefType t; NumType I32Type] --> [NumType I32Type]
+    let TableType (_lim, it, t) = table c x in
+    [RefType t; value_type_of_index_type it] --> [value_type_of_index_type it]
 
   | TableFill x ->
-    let TableType (_lim, _it, t) = table c x in
-    [NumType I32Type; RefType t; NumType I32Type] --> []
+    let TableType (_lim, it, t) = table c x in
+    [value_type_of_index_type it; RefType t; value_type_of_index_type it] --> []
 
   | TableCopy (x, y) ->
-    let TableType (_lim1, _it, t1) = table c x in
-    let TableType (_lim2, _it, t2) = table c y in
+    let TableType (_lim1, it1, t1) = table c x in
+    let TableType (_lim2, it2, t2) = table c y in
     require (t1 = t2) x.at
       ("type mismatch: source element type " ^ string_of_ref_type t1 ^
        " does not match destination element type " ^ string_of_ref_type t2);
-    [NumType I32Type; NumType I32Type; NumType I32Type] --> []
+    require (it1 = it2) x.at
+      ("type mismatch: source index type " ^ string_of_index_type it1 ^
+       " does not match destination index type " ^ string_of_index_type it2);
```

Comment:
I think this may not match memories where you can copy between 32/64-bit memories ok?

Reply:
The issue of copying between memories has not come up before since multi-memories have historically not been a thing.   In fact, the memory.copy instruction currently only take a single memory so its a non-issue there.   See #55.

Reply:
Oh I did bring this up [a few years back](https://github.com/WebAssembly/memory64/issues/6) and I implemented that issue's [proposed solution](https://github.com/WebAssembly/memory64/issues/6#issuecomment-668721640) in Wasmtime which has been there for some time now.

I don't have a strong motivation myself, but at least personally if table64 is added this doesn't seem really that much more of a stretch.